### PR TITLE
[ANE-1166] Support lock-file v6 for `pnpm`

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR, 

- Adds support for `pnpm` lock-file: `6.0`, which is default format with pnpm `8.x`.

Sibling PR: https://github.com/fossas/example-projects/pull/10

## Acceptance criteria

As a user, 
- [ ] I can analyze `pnpm` project using lock-file, with`6.0` 
- [ ] I can analyze `pnpm` workspace project using lock-file, with`6.0` 

## Testing plan

I added automated tests. 

- You can also test by performing `fossa analyze -o` against any `pnpm` project using `8.x`

For example, 
```bash
# Make fossa-cli
git pull origin && git checkout feat/pnpm-lock-v6 && make install-dev

# Get cases
cd ..
git clone -b feat/adds-pnpm-v6 git clone https://github.com/fossas/example-projects

# Analyze
fossa-dev analyze ./example-projects/javascript/pnpm-v6/ -o | jq
fossa-dev analyze ./example-projects/javascript/pnpm-v6-workspace/ -o | jq
```

Let's take popular project:
```bash
git clone https://github.com/vitejs/vite.git --depth 0
cd viite && fossa-dev analyze -o --only-target pnpm@./
```

There is also example: https://fossa.atlassian.net/browse/ANE-1166

## Risks

N/A

## Metrics

N/A

## References

https://fossa.atlassian.net/browse/ANE-1166

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).
- [x] If I made changes to a subcommand's options, I updated `docs/references/subcommands/<subcommand>.md`.

### CODE REVIEW

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 1]
```diff
@@ -76,17 +77,55 @@ import Path (Abs, File, Path)
...
+--    * Key of `packages` refer (e.g. "/buffer@4.9.2") denotes name of dependency and resolved version using '@' separator
+--        - For dependency resolved via registry resolver, format is: "/${dependencyName}@${resolvedVersion}${peerDepsInBrackets}".
+--      @
+--      >   /ieee754@1.1.13:
+--      >      resolution: {integrity: sha512...}
+--      >      dev: false
+--      >
+--      >   /@clerk/nextjs@4.22.1(next@13.4.10)(react-dom@18.2.0)(react@18.2.0):
+--      >       resolution: {integrity: sha512...}
+--      @
```
Comment (jssblck) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:102]: This says that peer dependencies are in brackets, but they are surrounded by parenthesis in the example. Which is correct? Please update accordingly.

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 2]
```diff
@@ -76,17 +77,55 @@ import Path (Abs, File, Path)
...
+--
+--    * If project has set peerDependencies to be not auto installed, pnpm
+--      by default, does not include them in the lockfile. So, no additional
+--      work is required for newly introduced `settings.autoInstallPeers` field.
```
Comment (jssblck) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:106]: I don't understand what this has to do with our usage. Can you clarify?

Specifically:
- Does FOSSA CLI need to do anything special to correctly report peer deps?
- If not, why not? If so, why so, and what does it need to do?

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 3]
```diff
@@ -76,17 +77,55 @@ import Path (Abs, File, Path)
...
 
+data PnpmLockFileVersion = PnpmLock5 | PnpmLock6 deriving (Show, Eq, Ord)
+
 instance FromJSON PnpmLockfile where
   parseJSON = Yaml.withObject "pnpm-lock content" $ \obj -> do
+    rawLockFileVersion :: TextLike <- obj .:? "lockfileVersion" .!= (TextLike mempty)
+    let lockFileVersion =
+          if Text.isPrefixOf "6" (unTextLike rawLockFileVersion)
+            then PnpmLock6
+            else PnpmLock5
```
Comment (jssblck) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:127]: This construction means that we'll parse lockfiles at v4 and below like v5, but it also means we'll parse lockfiles at v7 and above like v5 which seems incorrect.

Please compare the version in a way that properly accounts for this, unless there is a good reason for this functionality.

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 4]
```diff
@@ -121,6 +160,18 @@ instance FromJSON ProjectMap where
...
+  { version :: Text
+  }
+  deriving (Show, Eq, Ord)
+
+instance FromJSON ProjectMapDepMetadata where
+  -- This is v5 lock format
+  parseJSON (Yaml.String r) = pure $ ProjectMapDepMetadata r
+  -- This is v6 lock format
+  parseJSON (Yaml.Object obj) = ProjectMapDepMetadata <$> obj .: "version"
+  parseJSON _ = fail "Expected to contain string or object with 'version' field!"
```
Comment (jssblck) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:173]: Please print the thing that failed to parse properly to simplify later debugging.

e.g.:
```suggestion
  parseJSON other = fail ("Invalid format; expected pure string or an object with a `version` field, got: " <> show other)
```

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 5]
```diff
@@ -294,6 +372,14 @@ buildGraph lockFile = withoutLocalPackages $
     withoutSymConstraint :: Text -> Text
     withoutSymConstraint version = fst $ Text.breakOn "_" version
 
+    -- We do not care about peer dependency version
+    -- in the version postFix, for final transformation.
+    --
+    -- >> withoutPeerDepSuffix "1.2.0" = "1.2.0"
+    -- >> withoutPeerDepSuffix "1.2.0(babel@1.0.0)" = "1.2.0"
```
Comment (jssblck) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:379]: Please clarify why this is the case in the comment.

**Review by jssblck — CHANGES_REQUESTED** [review]
The v4/v7 handling is the only hard blocker here unless there's good reason for it working the way it does (or I'm misunderstanding).

File: src/Strategy/Node/Pnpm/PnpmLock.hs [hunk 6]
```diff
@@ -107,11 +149,17 @@ instance FromJSON PnpmLockfile where
             then Map.insert "." virtualRootWs importers
             else importers
 
-    pure $ PnpmLockfile refinedImporters packages
+    pure $ PnpmLockfile refinedImporters packages rawLockFileVersion
+    where
+      getVersion (TextLike ver) = case (listToMaybe . toString $ ver) of
```
Comment (meghfossa) [inline src/Strategy/Node/Pnpm/PnpmLock.hs:154]: We never supported version `<=3` previously, and there is no v7 as of yet.

Comment (jssblck) [reply src/Strategy/Node/Pnpm/PnpmLock.hs:154]: I think we should compare this as a number, and not die on v7 and above by default.
In other words, in the future v7 changes but in a way that doesn't render our parser incompatible, we shouldn't fail to parse it.

Comment (meghfossa) [reply src/Strategy/Node/Pnpm/PnpmLock.hs:161]: patched with: https://github.com/fossas/fossa-cli/pull/1320/commits/26204b4dfc6f6372c2c3885bb7c7fea77f1feadd

**Review by jssblck — APPROVED** [review]

**Pull request merged**